lib/pid/src/PID.cpp

```diff
@@ -1,492 +1,123 @@
-
-    // oscillate output
-    if (refVal > autoTuneConfig->setpoint + autoTuneConfig->noiseBand) {
-      setOutput(autoTuneConfig->outputStart - autoTuneConfig->oStep);
-    } else if (refVal < autoTuneConfig->setpoint - autoTuneConfig->noiseBand) {
-      setOutput(autoTuneConfig->outputStart + autoTuneConfig->oStep);
-    }
-
-    isMax = isMin = true;
-
-    // id peaks
-    for (int i = autoTuneConfig->nLookBack - 1; i >= 0; i--) {
-      float val = lastInputs[i];
-      if (isMax) isMax = refVal > val;
-      if (isMin) isMin = refVal < val;
-      lastInputs[i + 1] = lastInputs[i];
-    }
-    lastInputs[0] = refVal;
-
-    if (isMax) {
-      if (peakType == 0) peakType = 1;
-      if (peakType == -1) {
-        peakType    = 1;
-        justchanged = true;
-        peak2       = peak1;
-      }
-      peak1            = last_timer_time;
-      peaks[peakCount] = refVal;
-
-    } else if (isMin) {
-      if (peakType == 0) peakType = -1;
-      if (peakType == 1) {
-        peakType = -1;
-        peakCount++;
-        justchanged = true;
-      }
-
-      if (peakCount < 10) peaks[peakCount] = refVal;
-    }
-
-    // Transition, check if we can autotune based on the last peaks
-    if (justchanged && peakCount > 2) {
-      float avgSeparation =
-          (abs(peaks[peakCount - 1] - peaks[peakCount - 2]) + abs(peaks[peakCount - 2] - peaks[peakCount - 3])) / 2;
-      if (avgSeparation < 0.05 * (absMax - absMin)) break;
-    }
-    justchanged = false;
-
-    // wait for sample time interval
-    while ((timer.elapsed_time() - last_timer_time) < autoTuneConfig->sampleTime)
-      ;
-  }
-
-  // calculate autotune parameters TODO: FIXME! Clean this up. Where are these magic numbers from(@tandronescu)?
-  Ku = 4 * (2 * autoTuneConfig->oStep) / ((absMax - absMin) * M_PI);
-  Pu = (peak1 - peak2).count();
-
-  // if we only PI control is desired, alternate calculations and TauD set to 0
-  if (PI) {
-    autoTuneKc_   = 0.45 * Ku;
-    autoTuneTauR_ = 0.54 * Ku / Pu;
-    autoTuneTauD_ = 0;
-  } else {
-    autoTuneKc_   = 0.6 * Ku;
-    autoTuneTauR_ = 1.2 * Ku / Pu;
-    autoTuneTauD_ = 0.075 * Ku * Pu;
+float PID::Pid::computeDPathOnError(float error, int64_t dt) {
+  // no mutex lock needed since inside compute() only
+  float derivativePath = 0;
+  if (dt != 0) {
+    derivativePath = m_DGain * (error - m_pastError) / dt;
   }
+  return derivativePath;
 }
 
-void PID::setOutput(float output) {
-  // bound output to limits
-  if (output > outMax_)
-    output = outMax_;
-  else if (output < outMin_)
-    output = outMin_;
-
-  // cast void pointer based on required output type
-  // add desired output types here and convert from float as necessary
-  switch (actuatorType_) {
-    case PWM_MOTOR: {
-      *static_cast<PwmOut *>(output_) = output;
-      break;
-    }
-    default:
-      *(float *)output_ = output;
+float PID::Pid::computeDPathOnPV(float processVariable, int64_t dt) {
```

Comment:
need a sign flip somewhere because dError/dt = - dPV/dt. We don't want to need to switch the d gain from positive to negative when switching from antiKickback on vs off.

Reply:
good catch